integrations/octopus/main.py

```diff
@@ -0,0 +1,98 @@
+from enum import StrEnum
+from typing import Any, Dict
+from loguru import logger
+from port_ocean.context.ocean import ocean
+from port_ocean.core.ocean_types import ASYNC_GENERATOR_RESYNC_TYPE
+from client import OctopusClient
+
+TRACKED_EVENTS = [
+    "spaces",
+    "projects",
+    "deployments",
+    "releases",
+    "machines",
+]
+
+
+class ObjectKind(StrEnum):
+    SPACE = "space"
+    PROJECT = "project"
+    DEPLOYMENT = "deployment"
+    RELEASE = "release"
+    MACHINE = "machine"
+
+
+async def init_client() -> OctopusClient:
+    client = OctopusClient(
+        ocean.integration_config["octopus_api_key"],
+        ocean.integration_config["server_url"],
+    )
+    return client
+
+
+async def setup_application() -> None:
+    app_host = ocean.integration_config.get("app_host")
+    if not app_host:
+        logger.warning(
+            "App host was not provided, skipping webhook creation. "
+            "Without setting up the webhook, you will have to manually initiate resync to get updates from Octopus."
+        )
+        return
+    octopus_client = await init_client()
+    existing_subscriptions = await octopus_client.get_subscriptions()
+    existing_webhook_uris = {
+        subscription.get("EventNotificationSubscription", {}).get("WebhookURI")
+        for subscription in existing_subscriptions
+    }
+    webhook_uri = f"{app_host}/integration/webhook"
+    if webhook_uri in existing_webhook_uris:
+        logger.info(f"Webhook already exists with URI: {webhook_uri}")
+    else:
+        await octopus_client.create_subscription(app_host)
+        logger.info(f"Webhook created with URI: {webhook_uri}")
+
+
+@ocean.on_resync()
+async def on_global_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    octopus_client = await init_client()
+    async for resource_batch in octopus_client.get_paginated_resources(kind):
+        logger.info(f"Received length  {len(resource_batch)} of {kind} ")
+        yield resource_batch
+
+
+@ocean.router.post("/webhook")
+async def handle_webhook_request(data: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Handle the webhook request from Octopus Deploy.
+    """
+    payload = data.get("Payload", {}).get("Event", {})
+    related_document_ids = payload.get("RelatedDocumentIds", [])
+    event_category = payload.get("Category", "")
+    action = "unregister" if "Deleted" in event_category else "register"
```

Comment:
can you share a sample of the webhook event payload so we understand the logic below?

Reply:
This is a sample event for the project creation. It is generally same format with other events

`{
  "Timestamp": "2024-08-06T17:27:55.333+00:00",
  "EventType": "SubscriptionPayload",
  "Payload": {
    "ServerUri": "https://testport.octopus.app",
    "ServerAuditUri": "https://testport.octopus.app/app#/Spaces-1/configuration/audit?from=2024-08-06T17%3a27%3a24.%2b00%3a00&to=2024-08-06T17%3a27%3a54.%2b00%3a00",
    "BatchProcessingDate": "2024-08-06T17:27:54.670+00:00",
    "Subscription": {
      "Id": "Subscriptions-4",
      "Name": "Port Subscription",
      "Type": "Event",
      "IsDisabled": false,
      "EventNotificationSubscription": {
        "Filter": {
          "Users": [],
          "Projects": [],
          "ProjectGroups": [],
          "Environments": [],
          "EventGroups": [],
          "EventCategories": [],
          "EventAgents": [],
          "Tenants": [],
          "Tags": [],
          "DocumentTypes": []
        },
        "EmailTeams": [],
        "EmailFrequencyPeriod": "00:00:00",
        "EmailPriority": "Normal",
        "EmailDigestLastProcessed": null,
        "EmailDigestLastProcessedEventAutoId": null,
        "EmailShowDatesInTimeZoneId": null,
        "WebhookURI": "https://8bc2-102-89-22-116.ngrok-free.app/integration/webhook",
        "WebhookTeams": [],
        "WebhookTimeout": "00:00:50",
        "WebhookHeaderKey": null,
        "WebhookHeaderValue": null,
        "WebhookLastProcessed": "2024-08-06T17:27:24.568+00:00",
        "WebhookLastProcessedEventAutoId": 12157
      },
      "SpaceId": "Spaces-1",
      "Links": {
        "Self": "/api/Spaces-1/subscriptions/Subscriptions-4"
      }
    },
    "Event": {
      "Id": "Events-12290",
      "RelatedDocumentIds": [
        "Projects-3"
      ],
      "Category": "Created",
      "UserId": "Users-21",
      "Username": "explorer3107@gmail.com",
      "IsService": false,
      "IdentityEstablishedWith": "Session cookie",
      "UserAgent": "OctopusClient-js/2024.3.8427",
      "Occurred": "2024-08-06T17:27:38.970+00:00",
      "Message": "Project Newest Project was created",
      "MessageHtml": "Project <a href='#/projects/Projects-3'>Newest Project</a> was created",
      "MessageReferences": [
        {
          "ReferencedDocumentId": "Projects-3",
          "StartIndex": 8,
          "Length": 14
        }
      ],
      "Comments": null,
      "Details": null,
      "ChangeDetails": {
        "DocumentContext": {
          "SpaceId": "Spaces-1",
          "Id": "Projects-3",
          "Name": "Newest Project",
          "Slug": "newest-project",
          "Description": "Just another test",
          "ExecuteDeploymentsOnResilientPipeline": null,
          "IsDisabled": false,
          "VariableSetId": "variableset-Projects-3",
          "DeploymentProcessId": "deploymentprocess-Projects-3",
          "DeploymentSettingsId": "deploymentsettings-Projects-3",
          "ClonedFromProjectId": null,
          "ProjectGroupId": "ProjectGroups-3",
          "LifecycleId": "Lifecycles-3",
          "AutoCreateRelease": false,
          "PersistenceSettings": {
            "Type": "Database"
          },
          "DynamicEnvironmentSettings": {
            "ProvisioningRunbook": null,
            "DeprovisioningRunbook": null
          },
          "DiscreteChannelRelease": false,
          "IncludedLibraryVariableSetIds": [],
          "UsedPackages": [],
          "TenantedDeploymentMode": "Untenanted",
          "Templates": [],
          "ReleaseCreationStrategy": {
            "ReleaseCreationPackage": null,
            "ChannelId": null
          },
          "AutoDeployReleaseOverrides": [],
          "ExtensionSettings": [],
          "AllowIgnoreChannelRules": true,
          "DataVersion": null
        },
        "Differences": []
      },
      "IpAddress": "102.89.34.130",
      "SpaceId": "Spaces-1",
      "Links": {
        "Self": "/api/events/Events-12290"
      }
    },
    "BatchId": "3531f2c3-102a-4e18-9d6f-5dacf4d7ae58",
    "TotalEventsInBatch": 6,
    "EventNumberInBatch": 1
  }
}
`